src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,325 @@
+      //   In that case, we retry the WaitExecution call on the same operation object.
+      // - A WaitExecution call may fail with a NOT_FOUND error (raise a StatusRuntimeException).
+      //   That means the Operation was lost on the server, and we will retry to Execute.
+      // - Any call can return an Operation object with an error status in the result. Such
+      //   Operations are completed and failed; however, some of these errors may be retriable.
+      //   These errors should trigger a retry of the Execute call, resulting in a new Operation.
+      Preconditions.checkState(lastOperation == null);
+
+      ExecuteResponse response = null;
+      // Exit the loop as long as we get a response from either Execute() or WaitExecution().
+      while (response == null) {
+        // We use refreshIfUnauthenticated inside retry block. If use it outside, retrier will stop
+        // retrying when received a unauthenticated error, and propagate to refreshIfUnauthenticated
+        // which will then call retrier again. It will reset the retry time counter so we could
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                executeBackoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implemented
+        // inside the retry block.
+        //
+        // The cases to exit the loop:
+        //   1. Received the final response.
+        //   2. Received a un-retriable gRPC error.
+        //   3. Received NOT_FOUND error where we will retry Execute() (by returning null).
+        //   4. Received consecutive retriable gRPC errors (up to max retry times).
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              waitExecutionBackoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get().execute(request);
+        return handleOperationStream(operationStream);
+      } catch (StatusRuntimeException e) {
+        // If lastOperation is not null, we know the execution request is accepted by the server. In
+        // this case, we will fallback to WaitExecution() loop when the stream is broken.
+        if (lastOperation != null) {
+          // At this point we know that the waitExecutionBackoff is just reset, so call
+          // nextDelayMillis() here to ensure the expected retry times(--remote_retry + 1) for
+          // waitExecutionBackoff: once you reset it, the next call to nextDelayMillis() will not
+          // increase the counter.
+          waitExecutionBackoff.nextDelayMillis();
```

Comment:
opt: since this is only so that you get the exact number of retries, consider removing this for the sake of simplicity (note the opt tag -- I don't feel very strongly about this)

Reply:
Removed.